server.go

```diff
@@ -287,21 +322,73 @@ func (s *Server) runHandler(handler HandlerFunc, deliveries <-chan amqp.Delivery
 
 		ctx := context.WithValue(context.Background(), CtxQueueName, queueName)
 
-		// TODO: Recover panic in handler. And use delivery.Nack() so amqp will retry.
-		handler(ctx, &rw, delivery)
-		delivery.Ack(false)
+		// Process the request in a separate go routine to not crash the server
+		// on panic and make use of the panic recovery.
+		go func(delivery amqp.Delivery) {
+			pr := processedRequest{
+				replyTo:   delivery.ReplyTo,
+				mandatory: rw.mandatory,
+				immediate: rw.immediate,
+			}
 
-		s.responses <- processedRequest{
-			replyTo:    delivery.ReplyTo,
-			mandatory:  rw.mandatory,
-			immediate:  rw.immediate,
-			publishing: *rw.publishing,
-		}
+			defer s.handlerRecovery(delivery, rw, pr, crashes, &mu)
+
+			handler(ctx, &rw, delivery)
+			delivery.Ack(false)
+
+			pr.publishing = *rw.publishing
+			s.responses <- pr
+		}(delivery)
 	}
 
 	logger.Infof("server: stopped waiting for messages on queue '%s'", queueName)
 }
 
+func (s *Server) handlerRecovery(delivery amqp.Delivery, rw ResponseWriter, pr processedRequest, crashes map[string]float64, mu *sync.RWMutex) {
+	r := recover()
+	if r == nil {
+		return
+	}
+
+	logger.Warnf("server: crashed while handling request")
+
+	// Add new crash to the request recovering from a panic.
+	mu.RLock()
+	crashes[delivery.CorrelationId]++
+	mu.RUnlock()
+
+	switch {
+	case crashes[delivery.CorrelationId] > s.maxRetries:
+		fmt.Fprintf(&rw, "crashed when running handler: %s", r.(string))
+		rw.WriteHeader(ServerCrashedHeader, r.(string))
+
+		pr.publishing = *rw.publishing
+
+		s.responses <- pr
+
+		// Time to remove the requet from crash map, we could not avoid a panic
+		// within the max allowed retry attempts.
+		mu.Lock()
+		delete(crashes, delivery.CorrelationId)
+		mu.Unlock()
+
+		// Nack message and do NOT requeue.
+		delivery.Nack(true, false)
+	default:
+		// Use exponential f(x) = 2^x as backoff but never more than one hour.
+		backoff := math.Pow(2, crashes[delivery.CorrelationId])
+		if backoff > 3600 {
+			backoff = 3600
+		}
+
+		logger.Warnf("server: max retries not reached (%v/%v), will retry in %v seconds", crashes[delivery.CorrelationId], s.maxRetries, backoff)
+
+		time.Sleep(time.Duration(backoff) * time.Second)
```

Comment:
What happens when the server is stopped?

Reply:
Nack will deliver the message to a dead letter queue if such is configured.

Reply:
Wouldn’t a middleware make more sense? That way the user can implement his own if other responses/handling are needed.

Reply:
I guess a middleware could be a good idea here. You got anything on top of your mind or do you think this is enough?

```go
func PanicRecovery(next amqprpc.HandlerFunc) amqprcp.HandlerFunc {
    func(ctx context.Context, rw *ResponseWriter d amqp.Delivery) {
        go func() {
            defer func() {
                if r := recover(); r != nil {
                    // handle recover
                    next(ctx, rw, d)
                }
            }()
        }
    }
}

s := NewServer(url).WithMiddleware(PanicRecovery)
```

Btw: if it's a middleware, should we provide it in the framework, as some kind of "plugin" or just as an example to the end user?

Regarding what happens when the server is stopped I haven't really tested.

Reply:
Regarding `Nack()`, it'll be deliverd to the dead letter queue if `requeue` is false. The previous TODO comment was something like "[...] and requeue" but I thought that was wierd as well.

Reply:
I think the middleware you described would work perfectly. But what’s the reason for running it in a go routine?

Reply:
Oh, and what happens if we would Nack the delivery in the middleware? Because we Ack it later in the server.

Reply:
See https://github.com/bombsimon/amqp-rpc/issues/27

Reply:
The go routine was instead of the one in the server but will obv. be in the consumer :)

I have an almost finished example but I can't solv the issue where I avoid `Ack` on those panicked messages. And I would like a general solution either way if someone actually acks, nacks or rejects a message outside the server.

If I understand the [documentation](https://godoc.org/github.com/streadway/amqp#Channel.NotifyConfirm) I should be able to listen for confirms on two channels but nothing ever gets sent to them the way I've done it now.

```go
ackCh, nackCh := inputCh.NotifyConfirm(make(chan uint64), make(chan uint64))

// Start a routine that will listen on all ack/nack confirms and add
// the delivery tags that's being acked outside this loop.          
processedMessages := map[uint64]struct{}{}                          
                                                                    
go func() {                                                         
    for {                                                           
        select {                                                    
        case id := <-ackCh:                                         
            processedMessages[id] = struct{}{}                      
        case id := <-nackCh:                                        
            processedMessages[id] = struct{}{}                      
        }                                                           
    }                                                               
}()                                                                 
                                                                    
go func(delivery amqp.Delivery) {                                   
    handler(ctx, &rw, delivery)                                     
                                                                    
    // Ack the message if it's not already handled.                 
    if _, ok := processedMessages[delivery.DeliveryTag]; !ok {      
        delivery.Ack(false)                                         
    }                                                               
}()
```

Are you seeing the obvious error here? Am I misunderstanding the docs?

Reply:
Never mind, I'm on a fix!